src/core/api/registry.go

```diff
@@ -0,0 +1,201 @@
+// List lists all registries that match a given registry name.
+func (t *RegistryAPI) List() {
+	name := t.GetString("name")
+
+	_, registries, err := t.manager.List(&model.RegistryQuery{
+		Name: name,
+	})
+	if err != nil {
+		log.Errorf("failed to list registries %s: %v", name, err)
+		t.CustomAbort(http.StatusInternalServerError, http.StatusText(http.StatusInternalServerError))
+	}
+
+	// Hide passwords
+	for _, registry := range registries {
+		registry.Credential.AccessSecret = "*****"
+	}
+
+	t.Data["json"] = registries
+	t.ServeJSON()
+	return
+}
+
+// Post creates a registry
+func (t *RegistryAPI) Post() {
+	registry := &model.Registry{}
+	t.DecodeJSONReqAndValidate(registry)
+
+	reg, err := t.manager.GetByName(registry.Name)
+	if err != nil {
+		log.Errorf("failed to get registry %s: %v", registry.Name, err)
+		t.CustomAbort(http.StatusInternalServerError, http.StatusText(http.StatusInternalServerError))
+	}
+
+	if reg != nil {
+		t.HandleConflict(fmt.Sprintf("name '%s' is already used", registry.Name))
+		return
+	}
+
+	id, err := t.manager.Add(registry)
+	if err != nil {
+		log.Errorf("Add registry '%s' error: %v", registry.URL, err)
+		t.CustomAbort(http.StatusInternalServerError, http.StatusText(http.StatusInternalServerError))
+	}
+
+	registry.ID = id
+	t.Data["json"] = registry
+	t.ServeJSON()
+}
+
+// Put updates a registry
+func (t *RegistryAPI) Put() {
+	id := t.GetIDFromURL()
+
+	registry, err := t.manager.Get(id)
+	if err != nil {
+		log.Errorf("Get registry by id %d error: %v", id, err)
+		t.CustomAbort(http.StatusInternalServerError, http.StatusText(http.StatusInternalServerError))
+	}
+
+	req := models.RegistryUpdateRequest{}
+	t.DecodeJSONReq(&req)
+
+	originalName := registry.Name
+
+	if req.Name != nil {
+		registry.Name = *req.Name
+	}
+	if req.URL != nil {
+		registry.URL = *req.URL
+	}
+	if req.CredentialType != nil {
+		registry.Credential.Type = (model.CredentialType)(*req.CredentialType)
+	}
+	if req.AccessKey != nil {
+		registry.Credential.AccessKey = *req.AccessKey
+	}
+	if req.AccessSecret != nil {
+		registry.Credential.AccessSecret = *req.AccessSecret
+	}
+	if req.Insecure != nil {
+		registry.Insecure = *req.Insecure
+	}
+
+	t.Validate(registry)
+
+	if registry.Name != originalName {
+		reg, err := t.manager.GetByName(registry.Name)
+		if err != nil {
+			log.Errorf("Get registry by name '%s' error: %v", registry.Name, err)
+			t.CustomAbort(http.StatusInternalServerError, http.StatusText(http.StatusInternalServerError))
+		}
+
+		if reg != nil {
+			t.HandleConflict("name is already used")
+			return
+		}
+	}
+
+	if err := t.manager.Update(registry); err != nil {
+		log.Errorf("Update registry %d error: %v", id, err)
+		t.CustomAbort(http.StatusInternalServerError, http.StatusText(http.StatusInternalServerError))
+	}
+}
+
+// Delete deletes a registry
+func (t *RegistryAPI) Delete() {
+	id := t.GetIDFromURL()
+
+	_, err := t.manager.Get(id)
+	if err != nil {
+		if utilerr.Is(err, utilerr.ReasonNotFound) {
```

Comment:
I think this logic isn't needed anymore? and we need to handle 404 by checking the registry returned?

Reply:
My mistake